packages/playwright/src/util.ts

```diff
@@ -332,13 +346,33 @@ export function resolveImportSpecifierExtension(resolved: string): string | unde
   }
 
   if (dirExists(resolved)) {
+    const packageJsonPath = path.join(resolved, 'package.json');
+
+    if (isPathMapping) {
+      // After TypeScript path mapping, here's how directory imports are resolved
+      // with "moduleResolution:bundler" setting:
+      //  - `package.json#exports` is not respected, because it's not a proper module
+      //  - `package.json#main` is respected according to "folder as module" Node.js behavior
+      //  - `index.js` default is respected according to "folder as module" Node.js behavior
+      //
+      // More info:
+      //  - https://www.typescriptlang.org/docs/handbook/modules/reference.html#directory-modules-index-file-resolution
+      //  - https://nodejs.org/dist/latest-v20.x/docs/api/modules.html#folders-as-modules
+
+      const mainField = getMainFieldFromPackageJson(packageJsonPath);
+      const mainFieldResolved = mainField ? resolveImportSpecifierExtension(path.resolve(resolved, mainField), isPathMapping) : undefined;
```

Comment:
Why do we resolve package's main via path mapping? I don't think we should respect aliases in package.json, so should we just defer to Node from here?

Reply:
We do not apply path mappings here. `resolveImportSpecifierExtension` handles extension substitution, similarly to  [typescript](https://www.typescriptlang.org/docs/handbook/modules/reference.html#file-extension-substitution).

Deferring to Node does not work, that's exactly the issue we are fixing. TypeScript allows ESM directory imports, while pure Node does not.